document-portal/document-portal-fuse.c

```diff
@@ -2004,7 +2004,7 @@ xdp_fuse_read (fuse_req_t req,
   struct fuse_bufvec buf = FUSE_BUFVEC_INIT(size);
   XdpFile *file = (XdpFile *)fi->fh;
 
-  g_debug ("READ %lx size %ld off %ld", ino, size, off);
+  g_debug ("READ %" G_GINT64_MODIFIER "x size %" G_GSIZE_FORMAT " off %" G_GSIZE_FORMAT, ino, size, off);
```

Comment:
Sorry, I had missed that this commit is also correcting the format string for `size_t` and `off_t` in some places, not just `fuse_ino_t`.

For `size_t`, you are right that using `"%" G_GSIZE_FORMAT` or `"%zu"` is correct. I would personally suggest `"%zu"` for `size_t` in software that is not intended to be portable to non-Unix platforms, but `"%" G_GSIZE_FORMAT` is fine. (On Linux and all Unixes that I know of, `long` is the same size as `size_t` so `%ld` is equivalent to `%zd`, although strictly speaking wrong for `size_t` because of the different signedness.)

For `off_t`, unfortunately neither `"%ld"` nor `"%" G_GSIZE_FORMAT` is always correct. `off_t` is 64-bit on LP64 platforms, but can be either 32- or 64-bit on ILP32, depending on build options, toolchain, and which specific ILP32 platform (for example it's normally 32-bit on i386 but 64-bit on x32). Unfortunately there is no Standard C or POSIX modifier for an `off_t`.

I'm surprised that this compiled for you without warnings on a 32-bit platform, in fact, because I thought both Meson and FUSE enforce use of LFS (large file support), which makes `off_t` always 64-bit even on 32-bit platforms?

As far as I know, the only portable way to print an `off_t` would be with an explicit cast to a type that is at least as large as `off_t`: `("%lld", (long long) off)`, `("%jd", (intmax_t) off)`, or if you prefer the GLib type vocabulary, `("%" G_GOFFSET_FORMAT, (goffset) off)`.

Reply:
Indeed, the changes for `size_t` and `off_t` were not in our original patch, I just added them when preparing the branch for this upstream, and I tested them on x86-46 only. I'll try your suggestion (although quite verbose, using the GLib type seems cleaner to me) on 32 bits too.

Reply:
Done.

On arm32 I don't get any warnings on g_debug invocations, so this seems to be fine. I do get quite a few of these:

```
../document-portal/document-portal-fuse.c: In function 'xdp_fuse_setattr':
../document-portal/document-portal-fuse.c:1481:23: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
 1481 |       XdpFile *file = (XdpFile *)fi->fh;
      |                       ^
```
but this is harmless, since `fh` is declared as `uint64_t` and our pointers are 32 bits. I guess that an intermediate cast to `uintptr_t` would fix these warnings, if one really wanted to.

Reply:
This would be wrong on CHERI or other architectures with very large pointers, so strictly speaking this ought to have a `G_STATIC_ASSERT (sizeof (fi->fh) >= sizeof (void *))` or something, and then use `(XdpFile *) (uintptr_t) fi->fh` to silence the warning. But that seems more appropriate for a separate PR.

Reply:
Makes sense.